Bluetooth: BR: Bt classic auto pts setup

lylezhu2012 opened the pull request:
Due to the [PR](https://github.com/zephyrproject-rtos/zephyr/pull/74222) was closed due to the source branch has been deleted. Create a new PR for the changes.

This PR is used to add BR/EDR PTS support to bt_tester.

Test Suite for BR/EDR,
- [x] GAP
- [x] L2CAP
- [x] SMP
- [x] SDP

The changes for auto-pts script is [here](https://github.com/auto-pts/auto-pts/pull/1185)

lylezhu2012 committed:
include: bluetooth: Add definition for HCI_Write_Current_IAC_LAP

Add definition and structure for write current IAC LAP command

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Enable set_connectable for classic

Enable/disable connectable for classic when BTP
set_connectable command is received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Update set_discoverable for classic

Enable/disable discoverable for classic when BTP
set_discoverable command is received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Update start_discovery for classic

Start descovery for classic when BTP
start_discovery command is received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Update stop_discovery for classic

Stop discovery for classic when BTP
set_discoverable command is received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add gap_connect_br for classic

Add a command gap_connect_br for classic to create
br connection.

Add a structure btp_gap_connect_br_cmd for this
command.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Get conn from address for classic

Get the conn handle from the peer address for classic
when the following BTP command received, including
disconnect, pair, unpair, passkey_entry,
passkey_confirm, and conn_param_update.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support l2cap BREDR

Support L2CAP BREDR for BTP `listen` and
`connect` command.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Fix conn cannot be free issue

After the btp l2cap command `connect` called, the
conn cannot be free when the connect broken.

The function bt_conn_unref should be called to free
conn before the btp l2cap command `connect` done.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add device name to ext inquiry rsp

Add device name to extended inquiry response for
classic.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Correct peer addr for classic/spp.c

The peer address in BTP event is incorrect if the conn
type is classic.

Correct the peer address.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support pincode request

Add pincode entry if the IOCAP is
KeyboardDisplay or KeyboardOnly.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Host: SSP: Fix bonding type issue

Currently, the bonding type of Authentication
_Requirements parameter is always `Dedicated
Bonding` if the device is pairing initiator.

But if the bonding is performed during
connection setup or channel establishment as
a precursor to accessing a service, the
bonding type should be `General bonding`.

Add a flag BT_CONN_BR_GENERAL_BONDING.
Set the flag if the bonding is performed in
the L2CAP_BR/RFCOMM channel establishment.

Set bonding type depends on the flag when
receiving IO cap request.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Host: smp: Add function to get bonding setting

Add a function bt_get_bondable to get the bonding setting.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: SSP: Support non-binding case

If the return value of function
bt_get_bondable is false, clear the
bonding flag when controller requiring
`Authentication_Requirements`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add a GAP command pair_with_sec_level

In current implementation, the security level is a hard
code BT_SECURITY_L2.

Add GAP command pair_with_sec_level to set the security
level when starting pairing procedure.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add a L2CAP command connect_with_sec_level

For case `GAP/SEC/SEM/BV-09-C`, the first L2CAP connect
requires encryption with security level 2.
And for secondary L2CAP connect, it requires encryption
with security level 3.

For current implementation, the required security level
is always 2. It cannot meet the requirement.

Add a L2CAP command connect_with_sec_level to set the
security level when creating L2CAP connection.

In the command, pass the required security level to
field `required_sec_level` of L2CAP channel.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Custom ad flags

If the ad flags is set by auto-pts client,
overwrite the ad flags by using custom ad
flags.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: BREDR: L2CAP: Recovery ident when conn is pending

The `ident` of L2CAP BR connection req will be cleared if
function l2cap_br_conn_req_reply called to send L2CAP BR
connection rsp with result `BT_L2CAP_BR_PENDING`.

Then the invalid `ident` (it is zero) will be filled in
the L2CAP BR connection rsp after the ACL connection is
encrypted.

Recovery `ident` if the result of the connection rsp is
`BT_L2CAP_BR_PENDING`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: SSP: Support security level 4

Currently, error code `-ENOTSUP` will be
returned if start security with security
level 4.

For SC supported case, level 4 for ssp
should be supported.

Remove the code limitation to support
security level 4.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: SSP: Improve BR SC only mode

Actively disconnect the connection with
error code `BT_HCI_ERR_AUTH_FAIL` when
the notified link key type is not
`BT_LK_AUTH_COMBINATION_P256` in BR SC
only mode.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: BR: improve br_sufficient_key_size

Currently, the minimum value of encryption
key size is `BT_HCI_ENCRYPTION_KEY_SIZE_MIN`.

There is a Kconfig `BT_SMP_MIN_ENC_KEY_SIZE`.
It is used to set the specific minimum
encryption key size.

Use `CONFIG_BT_SMP_MIN_ENC_KEY_SIZE` as minimum
encryption key size in `br_sufficient_key_size`.

And update the `help` comments of
`BT_SMP_MIN_ENC_KEY_SIZE` to indicate the
configuration both for LE and classic.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: classic: Fix LE LTK cannot be derived issue

In the handler of SMP_Pairing rsp, the encryption key
flag is cleared incorrectly.

It causes the LE LTK cannot be derived.

Do not modify the encryption key flag to fix the
issue.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: l2cap_br: Support Multi-Command Packet

Improve L2CAP BR to handle more than one
signaling command in one receiving L2CAP packet.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Update l2cap_disconnect for classic

Find the correct l2cap channel according to channel
id.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Set flags in CFG RSP

The flags of L2CAP_CONFIGURATION_REQ are set.
But in L2CAP_CONFIGURATION_RSP, all bits of
flags are cleared.

When used in the L2CAP_CONFIGURATION_RSP
packet, the continuation flag shall be set
to one if the flag is set to one in the
Request.

Copy flags from Request to Response if it
is a successful result.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: incorrect result returned if config opt unsupported

The `BT_L2CAP_CONF_SUCCESS` is returned as result for the config
req with supported opt.

Result `BT_L2CAP_CONF_UNKNOWN_OPT` should be returned for this case.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Process all defined OPTs

Currently, Only configuration opt MTU is handled.

For opt `Flush timeout`, disconnect L2CAP channel
directly.

For opt `QOS`, `Retransmission and Flow Control`,
and `FCS`, response wilt result
`BT_L2CAP_CONF_UNACCEPT`.

For opt `extended flow specification` and
`extended windows size`, response wilt result
`BT_L2CAP_CONF_REJECT`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Update l2cap_send_data for classic

Find the correct l2cap channel according to channel
id.

And then use correct l2cap channel to send data.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Enable ECHO req/rsp feature

Enable ECHO req/rsp feature for BR.

Handle ECHO req/rsp packet.

And a function `bt_l2cap_br_echo` to send echo
req.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Tester: Add command l2cap_echo for BR

Add a l2cap command `echo` for BR.

Call function `bt_l2cap_br_echo` to send out echo
req.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add more options to connect command

Add following options to command L2CAP Connect,
`BTP_L2CAP_CONNECT_OPT_RET`
`BTP_L2CAP_CONNECT_OPT_FC`
`BTP_L2CAP_CONNECT_OPT_ERET`
`BTP_L2CAP_CONNECT_OPT_STREAM`

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Fix channel ops incorrect issue

Use `l2cap_br_ops` as ops of BR channel.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Tester: Support BR for command l2cap_credits

If no LE channel is in using, find channel in BR channels.

If the BR channel is found, call function
`bt_l2cap_chan_recv_complete` to free the buffer.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add l2cap_set_opt for classic

Add a command set_opt for classic to configure the
L2CAP channel.

Add a structure btp_l2cap_set_option_cmd for
this command.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support multiple pending credit for classic

Support L2CAP receiving multiple pending credit for classic.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: add a error code for unsupported case

Add error code `BTP_STATUS_NOT_SUPPORT`.

Return `BTP_STATUS_NOT_SUPPORT` if the error code of
function `bt_l2cap_chan_connect` is `-ENOTSUP`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add mode optional to listen_with_mode

Add opt `mode_optional` to command `listen_with_mode`.

When the option `mode_optional` is set, the field
`rx.optional` of L2CAP BR channel will be set when
the connect request is received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add mode optional to connect

Add option `mode_optional` to command `connect`.

When the option `mode_optional` is set, the field
`rx.optional` of L2CAP BR channel will be set when
sending connect request.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Tester: Add TX_BUFF_COUNT for L2CAP TX pool

Add a macro TX_BUFF_COUNT for L2CAP TX pool
And change the buffer count of TX pool from 2 to 4.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: send reject rsp if PDU is too small

Send reject response if the DPU length is less than
signaling header length, instead of doing nothing.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: SSP: clear pairing flag if ssp pairing completed

When the ssp pairing event is notified, clear the pairing
flag.

It is used to support case that the authentication is
started by peer, the link will not be encrypted after
the pairing is completed without any err.
If the local device want to encrypt the link, it could
call `bt_conn_set_security` to start encrypt the link
after the pairing complete callback triggered.

In original implementation, the `bt_conn_set_security`
cannot be called if the authentication has been started.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Tester: add mimxrt1170_evkb board specific setting

Add board overlay file,
* Use `dtcm` to `sdram`. This is the pin conflict between
M.2 pcm and sram.
* Set `zephyr,uart-pipe` to `&lpuart1`

Add board specific configuration,
* Select BT controller `CONFIG_BT_NXP_NW612`
* Enable BT Classic `CONFIG_BT_CLASSIC`
* Enable following configurations,
CONFIG_BT_L2CAP_MAX_WINDOW_SIZE=4
CONFIG_BT_L2CAP_ENH_RET=y
CONFIG_BT_L2CAP_STREAM=y
CONFIG_BT_L2CAP_FCS=y
CONFIG_BT_L2CAP_EXT_WIN_SIZE=y

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Sending multiple SDU at same time

Improve the retransmission and flow control to support
sending multiple SDU at the same time if the TX windows
is not full.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Add mode optional support

Add mode optional support for BR l2cap connect
initiator role.

If `chan->rx.optional` is true, set the mode
to basic mode instead of return error code
`-ENOTSUP`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: BR_SMP: Add a flag BT_CONN_BR_PAIRED

Due to the BT_CONN_BR_PAIRING is cleared when
receiving ssp pairing complete event. The LTK
key cannot be derived after the BR connection
encrypted.

Add a flag BT_CONN_BR_PAIRED that the pairing
has been done.

Use this flag to indicate whether the LTK
derivation needs to be applied if the BR ACL
connection is encrypted.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: BR: SMP: Check if remote supports CID 0x0007

Add a function bt_l2cap_br_get_remote_fixed_chan to
get the remote fixed channels.

If the fixed channel CID 0x0007 is unsupported, skip
the LTK derivation.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: BR: SMP: Set CT2 bit by default

Set CT2 bit by default.

If CT2 bit is set by both side, set the
CT2 flag `SMP_FLAG_CT2`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: LE: SMP: Set CT2 bit by default

Set CT2 bit to auth_req field by default if
the BR is enabled.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add service for SDP

Add service id BTP_SERVICE_ID_SDP=30 for SDP.

Implement function `tester_init_sdp` and
`tester_unregister_sdp`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Classic: SDP: Improve SDP discover

Extend the function `bt_sdp_discover` to support
service search transaction and service attribute
transaction.

Improve the `session->rec_buf`. If the net
buffer cannot be allocated from the channel,
disconnect the SDP session.

Set the `MaximumAttributeByteCount` of the
request `SDP_SERVICE_SEARCH_ATTR_REQ` with
the tail room of `session->rec_buf`.

Set the `MaximumAttributeByteCount` of the
request `SDP_SERVICE_ATTR_REQ` with
the tail room of `session->rec_buf`.

Set the `MaximumServiceRecordCount` of the
request `SDP_SERVICE_SEARCH_REQ` according to
the tail room of `session->rec_buf`.

Handle the response code
`SDP_SERVICE_SEARCH_RSP`, and
`SDP_SERVICE_ATTR_RSP`.

Handle the error `SDP_ERROR_RSP`. Start the
next SDP discovery if the error received.

If there no more request, disconnect the
session.

If the request cannot be sent, start the
next SDP discovery.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add a event for encryption failed

Add a event to report encryption failed if the
`err` of security callback `le_security_changed`
is not `BT_SECURITY_ERR_SUCCESS` or
BT_SECURITY_ERR_PIN_OR_KEY_MISSING.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: SDP discover

Add commands `BTP_SDP_SEARCH_REQ`,
`BTP_SDP_ATTR_REQ`, and
`BTP_SDP_SEARCH_ATTR_REQ`.

Add service record handle report event
`BTP_SDP_EV_SERVICE_RECORD_HANDLE`.

Handle the commands and call function
`bt_sdp_discover` to start the SDP
discovery.

And report the discovered service
record handle if the request type is
`BTP_SDP_SEARCH_REQ`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Shell: BR: Update command `sdp-find`

Since the function `bt_sdp_discover` has been
updated, the caller of function needs to be
updated to avoid the building and functionality
fault.

Add set the parameter `type` to value
`BT_SDP_DISCOVER_SERVICE_SEARCH_ATTR`.

Update the SDP discovery callback function.
Make it align with `bt_sdp_discover_func_t`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: HFP_AG: Update SDP discovery

Add set the parameter `type` to value
`BT_SDP_DISCOVER_SERVICE_SEARCH_ATTR`.

Update the SDP discovery callback function
by adding a third parameter
`const struct bt_sdp_discover_params *params`.
Make it align with `bt_sdp_discover_func_t`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
It doesn't seem to me like this needs to be in a public header file. I.e. I'd move it internal to `ssp.c` or just open code it in the single place you're using it (it really doesn't hurt readability IMO).

jhedberg commented on include/zephyr/bluetooth/hci_types.h:
Same with this one

jhedberg commented on subsys/bluetooth/host/classic/Kconfig:
What exactly do you need this feature for? I know it exists in the core spec, but I don't remember any real use cases for it. Let's not add things unless they're needed by some profile for which you have concrete product requirements.

jhedberg commented:
Please split this up into multiple smaller pull requests. It will be easier to review, and you'll avoid some frustration by being able to see quicker progress (some of the PRs will be possible to merge sooner than others).

lylezhu2012 committed:
Bluetooth: SSP: No MITM if required level is less than L3

Regardless IO capabilities, clear MITM flag for pairing
initiator if the required security level is less than
BT_SECURITY_L3.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: L2CAP_BR: Enable Retransmission and Flow control

Enable signaling channel configuration for Retransmission
and Flow control feature.

Send I-frame and S-frame. Support retransmission and Flow
control for sending.
Receive and handle I-frame and S-frame.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add listen_with_mode for classic

Add a command listen_with_mode for classic to listen
br L2CAP connection.

Add a structure btp_l2cap_listen_with_mode_cmd for
this command.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add change mode to opt for listen_with_mode

Use a option filed instead of mode to for BR l2cap listen.

Add a structure `struct bt_l2cap_br_server` to define the
L2CAP Server for BR. Add a filed `options` to save the opt
setting from command listen_with_mode.

Set the L2CAP connection `mode` and `hold_credit` in
callback `accept_br`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Add opt ext_win_size to listen_with_mode

Add opt `ext_win_size` to command `listen_with_mode`.

When the option `ext_win_size` is set, the field
`rx.extended_control` of L2CAP BR channel will be set
when the connect request is received.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on include/zephyr/bluetooth/hci_types.h:
Updated.
